### ISSUE TITLE
Consider moving to date-fns library

### ISSUE THREAD

**Issue by john2014** [fullcalendar/fullcalendar#3937]
In the upcoming version of full calendar, you mentioned removing jquery dependency.

Could you also consider moving from momentjs to date-fns? Here is the github link:

https://github.com/date-fns/date-fns

This library is far more efficient and smaller than momentjs and it has the same feature set. The file size is 8.5 KB (gzipped) compared to 50KB (gzipped) for momentjs. Due to smaller size, the browser also does not have to download and parse 50KB of data just to get few functions that full calendar required.

**Comment by acerix** [comment]
Thanks for the suggestion, sounds like quite an improvement, so I tagged this for further discussion.

**Comment by katie20** [comment]
I like the small file size of this new library with tons of features. Also since this change is internal to the library, it should not be considered as a breaking change (since nothing will break for the end user when you switch to the new date-fns library). Similarly it should not affect i18n (since it deals with internal date functions only and not any specific text that user will see).

**Comment by neroux** [comment]
No offence, but in which way is this even _remotely_ more "efficient"? **Please elaborate.**

The package consists as-is of ~600 files and clocks in at roughly 750 kilobytes. In comparison to that the single-file jQuery development(!) version seems lean at 260 kilobytes.

This seems to be yet another change for the sole sake of a change.

**Comment by john2014** [comment]
Filesize: Open your browser dev tool and load the library "https://cdnjs.cloudflare.com/ajax/libs/date-fns/1.29.0/date_fns.min.js". It is 10.70 KB in size (latest version). You can import specific module that you need.

Efficiency: https://raygun.com/blog/moment-js-vs-date-fns/ It is not only efficient for browser to parse and render but also efficient for users due to modular, easy function names, etc (see tutorial below).

Tutorial: https://www.sitepoint.com/date-fns-javascript-date-library/

I am also not a fan to change stuff for no reason. That time could be better spent on new features. But this is actually helpful. Plus other end users can easily reuse date-fns functions for their own code (if the library is included).

**Comment by neroux** [comment]
> Filesize

It is actually 70 KB. What you referred to is probably compressed and that is only relevant for the transfer, not for the parsing you mentioned.

Incidentally Moment.js is with 50 KB a third **smaller**.

> Efficiency

From a superficial glance it **appears** date-fns might perform better than Moment.js in some areas. This is in a benchmark scenario however (10,000+ loops) and I'd be careful to apply this as-is to fullcalendar.

> Tutorial

You didnt want to list this a pro, right?

> I am also not a fan to change stuff for no reason. That time could be better spent on new features. But this is actually helpful. Plus other end users can easily reuse date-fns functions for their own code (if the library is included).

I am sorry but I still miss the compelling reason for such a switch.

**Comment by john2014** [comment]
Every library has its pros and cons. It would be upto Dylan to make the final decision.

**Comment by neroux** [comment]
> Every library has its pros and cons.

Sure, but if there are no fundamental advantages there is little point it replacing an existing library with one that does not provide any substantial benefits.

**Comment by espen** [comment]
Related: the moment org released a new project recently: http://moment.github.io/luxon/

**Comment by frlinw** [comment]
In a perfect world, fullcalendar should not require any dependencies and the API should provide Date instances.
The user can chose to use a date lib or not (moment, date-dns, luxon, ... there is always a brand new thing).
It's the way to go to have a framework/lib agnostic fullcalendar (and it's for the best).

**Comment by neroux** [comment]
I'd second @frlinw's statement.

**Comment by arshaw** [comment]
I'm planning for v4 and made a proposal for overhauling the date system. Can you all take a look? Please make relevant comments on this thread.

[Wiki: Date System Revamp](https://github.com/fullcalendar/fullcalendar/wiki/Date-System-Revamp-Proposal)

**Comment by frlinw** [comment]
just for information, the latest trendy date lib : https://github.com/xx45/dayjs

**Comment by neroux** [comment]
> just for information, the latest trendy date lib : https://github.com/xx45/dayjs

Agreed, I'd generally suggest a more agile approach and re-evaluate the date library every couple of months ;)

**Comment by frlinw** [comment]
Just some questions about your post:
> However, we will still need translation files for generic strings like "Today", "Next", and "Prev" which appear in buttons of the calendar's header.
Can we consider the default header and its related stuff as a plugin in the v4 ? (removable on buildtime like SuperAgent)

> IE11 should still be supported. However, IE9/10 can be dropped.
What about a `next` branch targetting evergreen browsers ? too time-consuming/complex?

**Comment by arshaw** [comment]
1- that might be a good idea. i'll think about it. downside is that it adds complexity to a scenario that 90% of users need, and it's not actually that large of a filesize increase

2- a next branch might be too hard to maintain. will think about it after v4 is released

**Comment by frlinw** [comment]
1 - 90% of users need a header but it seems a lot of them need more control over it (dropdown instead of buttons list for responsive purpose, custom actions like show/hide weekend, ...)
I did a codepen for an other issue and I was surprised to see the number of views:
![image](https://user-images.githubusercontent.com/11896237/40323473-c01d4af6-5d35-11e8-87f4-d4c438d2839d.png)
Reduce the filesize is just a nice to have if it does not require extra work.

But about the size, Fullcalendar is quite large (the largest third party library in my case) so remove unused parts can help (Views for example, with explicit imports and declaration in options maybe) and remove legacy stuff too (next branch proposal) but I know it's not a top priority.

**Comment by john2014** [comment]
Another popular alternative to moment.js: https://github.com/iamkun/dayjs

**Comment by arshaw** [comment]
This issue has been addressed in the latest alpha release (4.0.0-alpha.2)
[See the blog post &raquo;](https://fullcalendar.io/blog/2018/10/release-v4-alpha2)

V4 now deals with native Date objects.

with a plugin for moment: https://fullcalendar.io/docs/v4/moment-plugins
and a plugin for luxon: https://fullcalendar.io/docs/v4/luxon-plugin

for the reasons why i think date-fns is a bad idea, see the bottom of this article:
https://fullcalendar.io/docs/v4/date-object

**Comment by frlinw** [comment]
[suggestion] use a fetch polyfill https://github.com/github/fetch (browser only) instead of superagent. You can just remove the polyfill when you'll drop ie11 support.

**Comment by Sparticuz** [comment]
date-fns has release two packages that may address the issues stated here. https://github.com/date-fns/utc and https://github.com/date-fns/tz

I might be willing to look into making a plugin for it. Is there any info documenting what needs to be done for a plugin @arshaw?
